Allow `command`, `byproducts`, and `environment` to be optionally set…

idunbarh opened the pull request:
… when using `in_toto.runlib.in_toto_record_start()` and `in_toto.runlib.in_toto_record_stop()`, fixes #531

Signed-off-by: Ian Dunbar-Hall <ian.dunbar-hall@lmco.com>

Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #531 **:

**Description of the changes being introduced by the pull request**:

This PR is to allow `command`, `byproducts`, and `environment` to be optionally set with using in_toto.runlib.in_toto_record_start()` and `in_toto.runlib.in_toto_record_stop()` to wrap commands.

Since these new arguments are optional, it should not affect existing users of the python package.

Additionally, I updated the `Link._validate_environment()` to validate that any dictionary keys can only be `variables`, `filesystem`, or `workdir`. Keys not on this list will cause an exception to be raised. 

Let me know if you think proactive enforcement of the spec for the `environment` section of metadata is not a good idea and I'll remove it.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

JustinCappos committed:
Include Purdue in acknowledgements

Signed-off-by: Justin Cappos <justincappos@gmail.com>

idunbarh commented:
@adityasaky when you have a moment, can you provide feedback on this PR? This is closes #531.

idunbarh commented:
@adityasaky looks like there were a couple linting issues. These were corrected and looking for the workflow to be approved a second time.

adityasaky commented on in_toto/models/link.py:
```diff
@@ -181,3 +181,9 @@ def _validate_environment(self):
       raise securesystemslib.exceptions.FormatError(
           "Invalid Link: field `environment` must be of type dict, got: {}"
           .format(type(self.environment)))
+    for key in self.environment.keys():
```
We should remove this, the specification doesn't mandate a format for `environment` and `byproducts`. https://github.com/in-toto/docs/blob/master/in-toto-spec.md#441-environment-information

idunbarh replied:
reverted to original and removed this validation.

adityasaky commented on in_toto/runlib.py:
```diff
@@ -708,13 +723,12 @@ def in_toto_record_start(step_name, material_list, signing_key=None,
 
   LOG.info("Creating preliminary link metadata...")
-  environment = {}
   if record_environment:
```
I think _one_ of `environment` and `record_environment` should be used. Any practical danger `record_environment` overwrites the `workdir` passed in explicitly using `environment`?

adityasaky replied:
cc @lukpueh

idunbarh replied:
Interested to hear @lukpueh's thoughts.

in the meantime, added argument validation which raises `securesystemslib.exceptions.FormatError` if both are set. Happy to revert depending on the discussion.

```
  if environment and record_environment:
    raise securesystemslib.exceptions.FormatError(
        "The arguments environment and record_environment are mutually"
        " exclusive. Please use one or the other.")
```

lukpueh replied:
IMO a passed `environment` should override the recorded environment. This should allow for the same use cases, and we wouldn't need any mutual exclusion handling.

idunbarh replied:
Sorry for the delay getting this PR updated.

Updated for override the recorded environment which `environment` is provided to `in_toto.runlib.in_toto_record_stop()` function.

adityasaky commented on in_toto/runlib.py:
```diff
@@ -784,6 +798,10 @@ def in_toto_record_stop(step_name, product_list, signing_key=None,
     base_path (optional): A path relative to which artifacts are recorded.
         Default is the current working directory.
 
+    byproducts (optional): A dictionary that lists byproducts of the link
```
Curious, at some point here, we stop needing in-toto-record, correct? All this information is available and could be plugged into a `Link` object manually as well?

idunbarh replied:
At least for me, the advantage of `in_toto.runlib.in_toto_record_start()` and `in_toto.runlib.in_toto_record_stop()` is handling creating and reading of the link files and key management. 

Sure ... its _not a ton_ of code, but I'd find myself copying 10's lines of `in_toto.runlib` code into the hoppr code base.

adityasaky replied:
Makes sense, thanks!

adityasaky reviewed (changes requested):
Some mostly nits, some discussion points, but overall this looks good to me. Thanks for the PR!

adityasaky commented on in_toto/runlib.py:
```diff
@@ -641,8 +641,20 @@ def in_toto_record_start(step_name, material_list, signing_key=None,
     base_path (optional): A path relative to which artifacts are recorded.
         Default is the current working directory.
 
+    command (optional): A list consisting of a command and arguments executed
```
I can see a slight case for this to go into `stop`, while we're pinging @lukpueh. But I don't think it's a big deal and we generally want to lean towards users knowing what command(s) they're running at `start` time. Definitely not something to block this PR!

lukpueh replied:
Why not move all new arguments to `stop`? This would make the change quite a bit smaller.

adityasaky reviewed (commented):
LGTM, aside from the open thread re `environment` and `record_environment` being used mutually. While we're at it, should `environment` also be added to `in_toto_run` do you think? It also currently only has `record_environment`.

PradyumnaKrishna committed:
Adds in-toto flavored GPG Key, Signer, and Signature

Adds in-toto flavored legacy GPGKey, GPGSigner, and GPGSignature
to be used with other Signer implementation added with the DSSE.

DSSE won't support this legacy GPG api but it is useful in
supporting old signature wrapper with gpg compatibilities, i.e.
`Metablock` to generate in-toto Metadata and signing with GPG.

This newer `GPGSigner` and `GPGKey` implementation provided by
securesystemslib utilizes updated metadata formats, that is
incompatible with current in-toto signature wrappers.

Hence, these `_LegacyGPGSigner` and `_LegacyGPGKey` implementation
is provided to continue old wrapper support without any breaking
change.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

idunbarh commented:
> LGTM, aside from the open thread re `environment` and `record_environment` being used mutually. While we're at it, should `environment` also be added to `in_toto_run` do you think? It also currently only has `record_environment`.

Happy to add this to `in_toto_run`.  Would you want the addition of environment available in the CLI too? 

My previous commits only expose this options if you are importing `in_toto.runlib` and not if you are using the CLI.

PradyumnaKrishna committed:
Add Legacy GPGSigner and GPGKey testcase

Adds testcase to validate the working of _LegacyGPGSigner,
_LegacyGPGKey and _LegacyGPGSignature.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

adityasaky commented:
> Would you want the addition of environment available in the CLI too?

I'm concerned that a full fledged environment flag would get clunky, would we have users pass in key value pairs? Reminded of docker's CLI for setting environment vars.

dependabot[bot] committed:
Bump pathspec from 0.10.3 to 0.11.0

Bumps [pathspec](https://github.com/cpburnz/python-pathspec) from 0.10.3 to 0.11.0.
- [Release notes](https://github.com/cpburnz/python-pathspec/releases)
- [Changelog](https://github.com/cpburnz/python-pathspec/blob/master/CHANGES.rst)
- [Commits](https://github.com/cpburnz/python-pathspec/compare/v0.10.3...v0.11.0)

---
updated-dependencies:
- dependency-name: pathspec
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

lukpueh commented on in_toto/runlib.py:
```diff
@@ -602,8 +602,8 @@ def in_toto_run(name, material_list, product_list, link_cmd_args,
-    exclude_patterns=None, base_path=None, record_environment=False,
-    normalize_line_endings=False, lstrip_paths=None):
+    exclude_patterns=None, base_path=None, command=None, environment=None,
+    record_environment=False, normalize_line_endings=False, lstrip_paths=None):
```
I think we should move new kwargs `command` and `environment` to the end, to not break existing users, who call this function without keywords.

Same goes for new kwarg in `in_toto_record_stop`.

idunbarh replied:
I moved all of the args to `in_toto.runlib.in_toto_record_stop()` and placed them at the end of the argument list.

lukpueh reviewed (changes requested):
Thanks for the patch, @idunbarh. It think this looks mostly good, but I'd love if we could extend the API as little as possible to support your use case.

I noted one blocker (order of arguments), and two suggestions (move all to `stop`, override recorded env) inline.

Commit:
Prepare v1.3.1

- Bump version
- Add changelog entry
- FIX: include tests in source distribution

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Fix linter warning

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Fix bandit linter invocation in tox.ini

We invoke bandit twice, once w/o util.py and once only for util.py,
using different sections. util.py was reoved in #402. So this case
handling is no longer relevant.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Prepare v1.3.1-1 Debian release

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

PradyumnaKrishna committed:
Raise exception in verify_signature method

_LegacyGPGKey verify_signature method returns True if
verification succeeds and False otherwise. This method lacks
information behind the failure in verification. Hence, the
method now returns None if verification passes, and raises an
exception accordingly for the verification failure.

* Tests updated accordingly.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

PradyumnaKrishna committed:
Refactor models.gpg api

models.gpg apis made non-public and moved to models._signer
All ``_LegacyGPGSignature``, ``_LegacyGPGSigner``, and
``_LegacyGPGKey`` are renamed ``GPGSignature``, ``GPGSigner`` and
``GPGKey`` respectively.

Tests are updated accordingly.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

dependabot[bot] committed:
Bump cryptography from 39.0.0 to 39.0.1

Bumps [cryptography](https://github.com/pyca/cryptography) from 39.0.0 to 39.0.1.
- [Release notes](https://github.com/pyca/cryptography/releases)
- [Changelog](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/39.0.0...39.0.1)

---
updated-dependencies:
- dependency-name: cryptography
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
Stop using securesystemslib.process

* Instead of `securesystemslib.process.run`, use `subprocess.run`
  directly. The former is only a thin wrapper over the latter,
  which provides no advantage in Py3-only times.

* Copy `securesystemslib.process.run_duplicate_streams` to a helper
  in runlib. This function is really only needed internally by
  in-toto-run tooling, for which it was developped originally
  (#160). It does not have to be public API in securesystemslib.
  Moreover, the function is not trivial and there have been
  multiple related bugs, one of them still unresolved
  (see secure-systems-lab/securesystemslib#337).

This patch allows us to deprecate `securesystemslib.process`, and
fix or workaround related issue wrt the specific runlib use case.

**Note to git historians:**
run_duplicate_streams was originally removed from in-toto in
in-toto/in-toto@29f063a, and now copied back from
https://github.com/secure-systems-lab/securesystemslib/blob/v0.26.0/securesystemslib/process.py

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

dependabot[bot] committed:
Bump cryptography from 39.0.1 to 39.0.2

Bumps [cryptography](https://github.com/pyca/cryptography) from 39.0.1 to 39.0.2.
- [Release notes](https://github.com/pyca/cryptography/releases)
- [Changelog](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/39.0.1...39.0.2)

---
updated-dependencies:
- dependency-name: cryptography
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

idunbarh committed:
Incorporating suggestions from @lukpueh in PR for #531.
Adding simple unittest for added in_toto_record_stop arguments.

Signed-off-by: Ian Dunbar-Hall <ian.dunbar-hall@lmco.com>

idunbarh commented:
> Thanks for the patch, @idunbarh. It think this looks mostly good, but I'd love if we could extend the API as little as possible to support your use case.
> 
> I noted one blocker (order of arguments), and two suggestions (move all to `stop`, override recorded env) inline.

@lukpueh great suggestion. Yeah, thats definitely simplifies moving everything to `in_toto.runlib.in_toto_record_stop()`. Updated the PR to have everything in `in_toto.runlib.in_toto_record_stop()` and add a unittest.

lukpueh commented on in_toto/runlib.py:
```diff
@@ -880,6 +881,22 @@ def in_toto_record_stop(step_name, product_list, signing_key=None,
+        "stdout" (str), "stderr" (str) and "return-value" (int).
+
+    environment (optional): A dictionary to capture information about
+        the environment to be added in the resulting link metadata eg.::
```
```suggestion
        the environment to be added in the resulting link metadata eg.::

```
Minor nit: sphincs requires a newline here, otherwise there's this error in the doc build log
```
docstring of in_toto.runlib.in_toto_record_stop:56: ERROR: Unexpected indentation.
```

Interestingly enough, the html build still finishes and the code block is rendered correctly. 🤷

lukpueh reviewed (changes requested):
Thanks for the updates! Please fix linting error and address inline comment, and we can merge this!

adityasaky reviewed (approved):
Aside from what's already been pointed out, LGTM. @idunbarh, could you rebase / squash and get rid of the merge commit? Thanks!
